Properly document the local and global locality attributes. (#13831)
coq/coq: doc/sphinx/language/core/modules.rst

See :ref:`command-line-options` for a more general view over the Coq command
line options.

.. _controlling-locality-of-commands:

Controlling the scope of commands with locality attributes
----------------------------------------------------------

> Many commands produce an effect that will persist in a given scope,
> which can typically be the one of the section or of the module in
> which the command was called. Locality :term:`attributes <attribute>`
> can be used to change the scope of this effect. Note that some
> commands, like :cmd:`Arguments` do not have the same default locality
> when they are used inside or outside sections.

Reviewer: ```suggestion
Many commands have effects that apply only within a specific scope,
typically the section or the module in
which the command was called. Locality :term:`attributes <attribute>`
can alter the scope of the effect. Note that some
commands, like :cmd:`Arguments` do not have the same default locality
when they are used inside or outside sections.
```

"typically"?  what other scopes are there other than sections and modules?
"Note that .." should be under `attr:: local`.  In this paragraph you could say
that a few commands don't follow the general pattern--that alerts the reader
that quirks lie ahead.

Author: > "typically"? what other scopes are there other than sections and modules?

Some commands have effects which persist module boundaries by default.

I've added an alert. Let me know what you think.

.. attr:: local

   In most cases, the :attr:`local` attribute limits the effect of the
   command to the current scope (section or module).

>    The only exception is when applied to :cmd:`Definition`,
>    :cmd:`Theorem` or variants. For these commands, the effect of the
>    :attr:`local` attribute will be that the defined object will not be
>    available through a shorter name when importing the module in which
>    it is defined.

Reviewer: ```suggestion
   The only exception is when :attr:`local` is applied to :cmd:`Definition`,
   :cmd:`Theorem` or their variants. Importing modules containing
   such commands makes the defined objects available only through
   their fully-qualified names rather than their unqualified names.
```

>    There is an alternative syntax for the :attr:`local` attribute
>    using the ``Local`` prefix (see :n:`@legacy_attr`).

Reviewer: ```suggestion
   The``Local`` prefix is an alternative syntax for the :attr:`local` attribute
   (see :n:`@legacy_attr`).
```

   .. note::

      For some commands, this is the only supported locality within
      sections (e.g., for :cmd:`Notation`, :cmd:`Ltac` and :ref:`Hint

>       <creating_hints>` commands).

Reviewer: "only supported locality"??  Not clear what is meant.  If you mean that Notation, etc don't allow global or export, say that more directly, and maybe under the description of those attributes

Author: These commands come with a locality (either the default one or the one set by the attribute). I think this explanation belongs here. It wouldn't make sense to repeat it for every other locality attribute, especially as we extend the list.

Reviewer: Perhaps "locality" should be "locality option" everywhere since "locality" is not a defined term.  The dictionary definition is "the position or site of something" though in CS it's used a bit more broadly.

Author: The more appropriate term should really be "scope", but my goal in this PR is not to revise the entire thing and locality is the term that has been used so far. In most places, I write "locality attribute" which is a tad more explicit.

.. attr:: export

>    The :attr:`export` attribute makes the effect of the command
>    persist when the section is closed and when the module containing
>    the command is imported.

Reviewer: ```suggestion
   The :attr:`export` attribute makes the effect of the command
   persist when the section is closed and applies the effect when the module containing
   the command is imported.
```

"persist" suggests that the effect is already present, not quite right for importing.

>    This attribute is supported in particular by the :ref:`Hint
>    <creating_hints>`, :cmd:`Set` and :cmd:`Unset` commands, although
>    for :ref:`Hint <creating_hints>` commands, it is not supported
>    inside sections.

Reviewer: ```suggestion
   This attribute is supported in particular by the :ref:`Hint
   <creating_hints>`, :cmd:`Set` and :cmd:`Unset` commands, although
   for :ref:`Hint <creating_hints>` commands, it is not supported
   inside sections.
```

is "in particular" needed?  There are others that are not important.

Author: The goal was to prevent the documentation to become incorrect as the list of commands supporting the attribute would expand.

Reviewer: You could say "Commands supporting this attribute include ...".  "In particular" implies that the listed items are somehow more important that those that are not listed.

.. attr:: global

>    In most cases, the :attr:`global` attribute makes the effect of the
>    command persist even when the current section or module is closed
>    and in other files as soon as the one containing the command is
>    (possibly transitively) loaded.
> 
>    For a few commands (like :cmd:`Notation` and :cmd:`Ltac`), this
>    attribute behaves as in the description of :attr:`export`.

Reviewer: ```suggestion
   In most cases, the :attr:`global` attribute makes the effect of the
   command persist even when the current section or module is closed.
   Loading the module (possibly transitively) applies the effect of the command.

   For a few commands (like :cmd:`Notation` and :cmd:`Ltac`), this
   attribute behaves like :attr:`export`.
```

Maybe leave out "In most cases," and just have a label "Exceptions" here and in a couple other spots?
Not clear to me whether the exceptions are well thought out or more like historical accidents.

Author: The exceptions are the result of history, but we should get rid of them eventually.

Author: See #13153 for the exception to the default semantics of `local` and #13833 for the exceptions to the default semantics of `global`.

Reviewer: While I don't want to become expert on all the variations, it would be nice to get to a point where there is a 1-1 relationship between attributes and behaviors.  For example, if superglobal behavior remains available for some commands, we might create a `superglobal` attribute to distinguish it from regular global behavior.  Such commands could also give an error message like "'global' is not supported, did you mean 'superglobal'"?

Author: The most common semantics of `global` is the superglobal one, so the idea is rather to eventually make sure this is the only semantics behind `global` (and to reduce the number of commands which support it).

>    There is an alternative syntax for the :attr:`global` attribute
>    using the ``Global`` prefix (see :n:`@legacy_attr`).

Reviewer: ```suggestion
   The ``Global`` prefix is an alternative syntax for the :attr:`global` attribute
   (see :n:`@legacy_attr`).
```

   .. warning::

>       Using the :attr:`global` locality is strongly discouraged
>       because the transitive nature of file loading gives little
>       control to the user. Prefer the use of the :attr:`export`
>       locality attribute when it is supported.

Reviewer: ```suggestion
      We strongly discourage using :attr:`global` in commands
      because the transitive nature of file loading gives the user little
      control. We recommend using the :attr:`export`
      locality attribute where it is supported.
```

See :ref:`command-line-options` for a more general view over the Coq command
line options.

.. _controlling-locality-of-commands:

Controlling the scope of commands with locality attributes
----------------------------------------------------------

Many commands have effects that apply only within a specific scope,
typically the section or the module in which the command was
called. Locality :term:`attributes <attribute>` can alter the scope of
the effect. Below, we give the semantics of each locality attribute

> but also alert that a few commands give a different meaning to the
> :attr:`local` and :attr:`global` attributes.

Reviewer: ```suggestion
while noting a few exceptional commands in which 
:attr:`local` and :attr:`global` attributes are interpreted differently.
```

.. attr:: local

   The :attr:`local` attribute limits the effect of the command to the
   current scope (section or module).

>    The ``Local`` prefix is an alternative syntax for the :attr:`local`

Reviewer: ```suggestion
   The ``Local`` :n:`@legacy_attr` is an alternative syntax for the :attr:`local`
```

or "legacy attribute" if that can be made to hyperlink

Author: I prefer to keep as is.

   attribute (see :n:`@legacy_attr`).

   .. note::

>       - For some commands, this is the only supported locality within

Reviewer: ```suggestion
      - For some commands, this is the only locality supported within
```

        sections (e.g., for :cmd:`Notation`, :cmd:`Ltac` and
        :ref:`Hint <creating_hints>` commands).

>       - For some other commands, this is the default locality within
>         sections even though other locality attributes are supported
>         as well (e.g., it is the case for the :cmd:`Arguments`
>         command).

Reviewer: ```suggestion
      - For other commands, this is the default locality within
        sections even though other locality attributes are supported
        as well (e.g., for the :cmd:`Arguments` command).
```

Of the two cases, "the only supported locality" and "the default locality", which is more more common?  If there's a definite preference for one or the other, maybe say "For _most_ commands"

Author: The two cases are not the only options (which seems implied by the "For other commands"). I've changed it into "For some" for more symmetry. There's no definite preference and the first case (only supported locality) is a limitation rather than a feature.

   .. warning::

      **Exception:** when :attr:`local` is applied to

>       :cmd:`Definition`, :cmd:`Theorem` or their variants, its
>       semantics is different: it makes the defined objects available
>       only through their fully-qualified names rather than their
>       unqualified names after an :cmd:`Import`.

Reviewer: ```suggestion
      :cmd:`Definition`, :cmd:`Theorem` or their variants, its
      semantics are different: when these objects are :cmd:`Import`\ed,
      they must be referenced with their fully-qualified names rather than their
      unqualified names.
```

Author: I've applied the "is" -> "are" change but not the rest because I don't want to lose the idea of the effect of `local` that is carried by "it makes".

.. attr:: export

   The :attr:`export` attribute makes the effect of the command
   persist when the section is closed and applies the effect when the
   module containing the command is imported.

>    This attribute is supported in particular by the :ref:`Hint
>    <creating_hints>`, :cmd:`Set` and :cmd:`Unset` commands, although
>    for :ref:`Hint <creating_hints>` commands, it is not supported
>    inside sections.

Reviewer: ```suggestion
   Commands supporting this attribute include :cmd:`Set`, :cmd:`Unset` and
   the :ref:`Hint <creating_hints>` commands, although the latter doesn't support it
   within sections.
```

.. attr:: global

   The :attr:`global` attribute makes the effect of the command
   persist even when the current section or module is closed.  Loading
   the file containing the command (possibly transitively) applies the
   effect of the command.

   The ``Global`` prefix is an alternative syntax for the
   :attr:`global` attribute (see :n:`@legacy_attr`).

   .. warning::

      **Exception:** for a few commands (like :cmd:`Notation` and

>       :cmd:`Ltac`), this attribute behaves like :attr:`export`.

Reviewer 2: This sounds like they also support `export`, but they don't.

Author: In the previous version of the patch I wrote "behaves as in the description of `export`" but it seemed a bit wordy.

   .. warning::

      We strongly discourage using the :attr:`global` locality
      attribute because the transitive nature of file loading gives
      the user little control. We recommend using the :attr:`export`

>       locality attribute where it is supported.

Reviewer: The idea is to eventually phase out "global"?

Author: I cannot say for sure if it will be ever deprecated but yes.
